pkg/pipeline/pipeline_project_integration.go

```diff
@@ -0,0 +1,199 @@
+	},
+	"environments": {
+		Type:     schema.TypeSet,
+		Optional: true,
+		Elem: &schema.Schema{
+			Type: schema.TypeString,
+		},
+		Description: "In a project, an array of environment names in which this pipeline source will be.",
+	},
+	"is_internal": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "Set this as false to create a Pipelines integration.",
+	},
+}
+
+func unpackProject(data *util.ResourceData, key string) ProjectJSON {
+	var project ProjectJSON
+	input := data.Get(key).(map[string]interface{})
+	project.Key = input["key"].(string)
+	project.Name = input["name"].(string)
+
+	return project
+}
+
+func packProject(d *schema.ResourceData, schemaKey string, project ProjectJSON) []error {
+	var errors []error
+	log.Println("[DEBUG] packProject", project)
+	if (ProjectJSON{}) == project {
+		return errors
+	}
+	setValue := util.MkLens(d)
+	errors = append(errors, setValue(schemaKey, project)...)
+	return errors
+}
+
+func unpackProjectIntegration(data *schema.ResourceData, formJsonValues []FormJSONValues) (ProjectIntegration, error) {
+	d := &util.ResourceData{data}
+
+	projectIntegration := ProjectIntegration{
+		Name:                  d.GetString("name", false),
+		ProjectId:             d.GetInt("project_id", false),
+		MasterIntegrationId:   d.GetInt("master_integration_id", false),
+		MasterIntegrationName: d.GetString("master_integration_name", false),
+		Environments:          d.GetSet("environments"),
+		IsInternal:            d.GetBool("is_internal", false),
+		Project:               unpackProject(d, "project"),
+		FormJSONValues:        formJsonValues,
+	}
+	return projectIntegration, nil
+}
+
+func packProjectIntegration(d *schema.ResourceData, projectIntegration ProjectIntegration) []error {
+	var errors []error
+	setValue := util.MkLens(d)
+
+	errors = setValue("name", projectIntegration.Name)
+	errors = append(errors, setValue("master_integration_id", projectIntegration.MasterIntegrationId)...)
+	errors = append(errors, setValue("master_integration_name", projectIntegration.MasterIntegrationName)...)
+	errors = append(errors, setValue("environments", projectIntegration.Environments)...)
+	errors = append(errors, setValue("is_internal", projectIntegration.IsInternal)...)
+	errors = append(errors, packProject(d, "project", projectIntegration.Project)...)
+
+	return errors
+}
+
+func readProjectIntegration(data *schema.ResourceData, m interface{}) ([]FormJSONValues, error) {
+	log.Printf("[DEBUG] readProjectIntegration")
+	projectIntegration := ProjectIntegration{}
+	resp, err := m.(*resty.Client).R().
+		SetResult(&projectIntegration).
+		Get(projectIntegrationsUrl + "/" + data.Id())
+	log.Println("[DEBUG] projectIntegration body: ", string(json.RawMessage(resp.Body())))
+	if err != nil {
+		return nil, err
+	}
+	log.Println("[DEBUG] projectIntegration Obj: ", projectIntegration)
+	errors := packProjectIntegration(data, projectIntegration)
+	if len(errors) > 0 {
+		return nil, errors[0]
```

Comment:
your condition checks for `>0` - but what about `infinity -1` other errors? I mean, in that function it makes numerous calls that might produce errors. There is code everywhere in the other providers that handles this. So, this code simply is insufficient